### DISCUSSION TITLE
NLP sensitivity, higher-order derivatives and Jacobian-times-vector products

### DISCUSSION THREAD

**Discussion by MHW-UDS** [casadi/casadi#3267] [Q&A]
I'm interested in computing higher-order derivatives of functions and/or solver instances. In the following I will refer to "Sensitivity Analysis for Nonlinear Programming in CasADi" by @jaeandersson and Rawlings when using names for quantities and equation numbers. Suppose I want to obtain sensitivity information about the  optimization variable x w.r.t. p. This would be possible by evaluating eq. (10) for z:=x and q:=p. 

- Since (for now) I'm interested in dG/dq (i.e. computing jacobian(...)) and not \hat{z} = dG/dq \hat{q} (i.e. solvinge jtimes(F,q,\hat{q}) + jtimes(F,z,\hat{z}) = 0?), I would need to compute the full, potentially expensive, Jacobian, right? 
- Computing the (matricezed version of the) jacobian of dG/dq would then even be much more expensive? 
- Do you plan to implement a Taylor AD mode, s.t. common evaluations are shared and efficiency is improved?

Thanks

**Comment by jaeandersson** [comment]
> Since (for now) I'm interested in dG/dq (i.e. computing jacobian(...)) and not \hat{z} = dG/dq \hat{q} (i.e. solvinge jtimes(F,q,\hat{q}) + jtimes(F,z,\hat{z}) = 0?), I would need to compute the full, potentially expensive, Jacobian, right?

Computing the full Jacobian should not be very expensive. The expensive part of the solution is to form and factorize the linear system of equations, i.e. the linearized KKT system. But this factorization is something you need to do as part of solving the NLP anyway, so the sensitivity calculations does not really involve any additional calculations. In practice it might anyway, since CasADi does not typically have access to the factorized KKT systems from the interfaced solvers, but that's an implementation issue (that could potentially be worked around).

> Computing the (matricezed version of the) jacobian of dG/dq would then even be much more expensive?

If you need the full Jacobian, you are typically best off just asking for that information to be calculated.

> Do you plan to implement a Taylor AD mode, s.t. common evaluations are shared and efficiency is improved?

No, not unless it's needed by some project. There might also be easier ways to get the same speedup, like caching the calculations.
